### PR TITLE
[API] Improve metrics, documentation, linting

### PR DESCRIPTION
**Pull request by banool**
## Summary
This PR makes various improvements to metrics (per-endpoint logging), docs, spec generation (a new standalone binary), deduplication, and linting (to ensure spec changes are checked in).

I can make the failpoint stuff middleware as well once https://github.com/poem-web/poem/issues/351 goes in.

## Test plan
```
cargo nextest run --package smoke-test --test-threads 8
```

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2317)

### CODE REVIEW

File: .github/workflows/api-spec-lint.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+# Cancel redundant builds
+concurrency:
+  # Cancel redundant builds on PRs (only on PR, not on branches)
+  group: ${{ github.workflow }}-${{ (github.event_name == 'pull_request' && github.ref) || github.sha }}
+  cancel-in-progress: true
+
+jobs:
+  # Confirm that any changes to the API specs have been checked in.
+  api-spec-check:
```
Comment (geekflyer) [inline .github/workflows/api-spec-lint.yaml:24]: honestly, it'd just add this job to the existing `lint-test.yaml` workflow.

File: Cargo.toml [hunk 2]
```diff
@@ -3,6 +3,7 @@ resolver = "2"
 
 members = [
     "api",
+    "api/spec-generator",
```
Comment (geekflyer) [inline Cargo.toml:6]: mind renaming this to `openapi-spec-generator` for clarity?

File: api/spec-generator/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "aptos-api-spec-generator"
```
Comment (geekflyer) [inline api/spec-generator/Cargo.toml:2]: aptos-openapi-spec-generator

File: api/spec-generator/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+version = "0.1.0"
+authors = ["Aptos Labs <opensource@aptoslabs.com>"]
+description = "Aptos API OpenAPI Spec Generator"
+repository = "https://github.com/aptos-labs/aptos-core"
+homepage = "https://aptoslabs.com"
+license = "Apache-2.0"
+publish = false
+edition = "2018"
```
Comment (geekflyer) [inline api/spec-generator/Cargo.toml:10]: would recommend to put higher edition (I think we have 2021 or so somewhere in the root of the repo)

**Review by geekflyer — APPROVED** [review]
lgtm although I'm a Rust noob ;-)

File: api/spec-generator/Cargo.toml (hunk 4 above)
Comment (banool) [reply api/spec-generator/Cargo.toml:10]: Seems like 2018 is still the standard:
```
$ rg 'edition = ' | grep 2021 | wc -l
       2
$ rg 'edition = ' | grep 2018 | wc -l
     124
```

File: .github/workflows/api-spec-lint.yaml (hunk 1 above)
Comment (banool) [reply .github/workflows/api-spec-lint.yaml:24]: I believe Wolfgang is adding another similar check, perhaps we could have a "ensure-changes-checked-in.yaml" kind of thing?

**Comment by banool** [comment]
@rajkaramchedu, for when you look at this week, a quick guide. With the new API, the spec is generated from the code, vs them being separately written by hand. As such, all the documentation, examples, etc. exist alongside the relevant code (the endpoint handlers, the types, etc).

Some relevant paths include:
- `api/src/poem_backend/`, particularly the endpoint files, like `events.rs`, `transactions.rs`, etc. In those you'll see large comments above functions like `submit_transaction`; those are what end up in the spec.
- `api/types/src/`, particularly types in `derives.rs`.

If you want to play around with it, the flow is:
1. Change code, e.g. the comments above an endpoint handler function
2. Run the following command from the root of the repo to get a spec out:

```
cargo run -p aptos-openapi-spec-generator -- -f yaml -o api/doc/v1/spec.yaml
```

**Comment by banool** [comment]
@rajkaramchedu I keep running into merge conflicts because things keep changing around this PR, and even more so its child PR here: https://github.com/aptos-labs/aptos-core/pull/2333. I'm going to land this and we can make any docs changes in another PR.

**Pull request merged**